src/engine/server_iv.c

```diff
@@ -509,6 +509,9 @@ ivc_on_refresh(crt_iv_namespace_t ivns, crt_iv_key_t *iv_key,
 	       crt_iv_ver_t iv_ver, d_sg_list_t *iv_value,
 	       bool invalidate, int refresh_rc, void *priv)
 {
+	if (refresh_rc != 0)
+		return refresh_rc;
+
```

Comment:
Hmm, I don't think it should return refresh_rc here but instead SUCCESS. callback implementation's return indicates the status of its execution, i dont think you should just return iv framework the same rc it gave you for this value.

Reply:
hmm, refresh_rc is the result of remote fetch, if it failed, then it suppose to return skip the refresh. I thought  we should return the failure if refresh does not succeed here. since refresh_rc is one of the input anyway?

Reply:
no, not really. ivo_on_refresh() is just a notification to the IV member that the value is being updated. If IV value is unavailable/invalid then its still a 'valid' update, even if you chose to ignore it in your case. 
As far as IV goes ivo_on_refresh() errors should not affect any logic. So I think the right way would be to return DER_SUCCESS if refresh_rc != 0, and keep iv logic as it is today

Reply:
The concern is that you are changing IV behavior where previously ivo_on_refresh() errors were not meant to affect the fetch path (they totally should fail the update path) but now they do. I don't have issues if you want to change ivo_on_refresh() semantics to mean that any error should stop the fetch process and return appropriate on_refresh() error, but then this is incomplete and needs to address all other instances where ivo_on_refresh() is called.

In your particular case, if the fetch itself was successful, but for some reason local ivo_on_refresh() failed on the current node in the middle, then it will cause the wrong order of pending events finalization (as it would assume overall fetch failed).

Reply:
>but then this is incomplete and needs to address all other instances where ivo_on_refresh() is called.

Not sure I follow here, but some of ivo_on_fresh() return values are being checked in crt_iv.c

> In your particular case, if the fetch itself was successful, but for some reason local ivo_on_refresh() failed on the current node in the middle, then it will cause the wrong order of pending events finalization (as it would assume overall fetch failed).

As I understand,  if IV fetch failed in the middle of the path (after get the value from the root), those nodes already got the value(before failure) will keep those value anyway (by setting entry valid),   those nodes (after failure)  will be notified by failure anyway.  I do not understand why it has to be succeed along the path, how it is different with network failure. or I miss sth?